### ISSUE TITLE
`users.users.<name>.autoSubUidGidRange = true` creates overlapping Ids for normal and system users

### ISSUE THREAD

**Issue by ppenguin** [NixOS/nixpkgs#259588]
### Describe the bug

Setting `users.users.<name>.autoSubUidGidRange = true` for a few normal and a few system users creates overlapping Ids, where it appears that the "counter" starts at the beginning for the system and normal user category.

### Steps To Reproduce
1. Set `users.users.<name>.autoSubUidGidRange = true` for a few normal users
2. Set `users.users.<name>.autoSubUidGidRange = true` for a few system users
3. Observe overlapping IDs in `/etc/sub*id`

### Expected behavior

ID ranges count linearly to avoid overlap, considering both system and normal users the same.

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 6.5.5-zen1, NixOS, 23.11 (Tapir), 23.11.20231004.7131f3c`
 - multi-user?: `yes`
 - sandbox: `relaxed`
 - version: `nix-env (Nix) 2.17.0`
 - nixpkgs: `/nix/store/hcdnrkjkrm30lg62aakdiah076iwjxhl-9gzb6bb0ghi5njm63sghyl7ia47n1ncn-source`
```

**Comment by ppenguin** [comment]
I noticed that here (https://github.com/NixOS/nixpkgs/blob/81e8f48ebdecf07aab321182011b067aafc78896/nixos/modules/config/users-groups.nix#L355C27-L355C27) the default for normal users is to generate sub*ids if none are specified and not explicitly disabled, which is ok (most system users wouldn't need to run podman), but I coundn't find where the actual generation script resides...

**Comment by WinstonHartnett** [comment]
Yep, ran into this too. The bit that updates user [info is here](https://github.com/NixOS/nixpkgs/blob/81e8f48ebdecf07aab321182011b067aafc78896/nixos/modules/config/users-groups.nix#L653C1-L663C1) by calling [this perl script](https://github.com/NixOS/nixpkgs/blob/440e21d286d33129f3eb8da73eaff46480d1289a/nixos/modules/config/update-users-groups.pl).

**Comment by agowa** [comment]
One suggestion, how about changing the allocation algorithm to a much more simpler one?

As autoSubUidGidRange is listed with a default of "false" over at search.nixos.org I initially configured my system the usual way I knew from other places like when e.g. an LDAP directory is being used with:
```sh
    subGidRanges = [
      {
        count = 65536;
        # Gid 0 == root system. So in order for the group "root" to have their own subGidRange
        # that doesn't overlay with the systems primary Gid range we add 1 to the gid.
        # then we multiply it with 65536 to allow each group to have their own range.
        # With systemd-homed the allocatable range is 524288-1878982656 (0x80000-0x6fff0000).
        startGid = (config.users.groups.${config.users.users.user.group}.gid + 1) * 65536;
      }
    ];
    subUidRanges = [
      {
        count = 65536;
        # Uid 0 == root system. So in order for the user "root" to have their own subUidRange
        # that doesn't overlay with the systems primary Uid range we add 1 to the gid.
        # then we multiply it with 65536 to allow each group to have their own range.
        # With systemd-homed the allocatable range is 524288-1878982656 (0x80000-0x6fff0000).
        startUid = (config.users.users.user.uid + 1) * 65536;
      }
    ];
```

This should adequately solve the duplicate ID issue reported in here. But I'm not quite sure why the choice for the current one was made and as it looks a bit more sophisticated there may be some reasons I'm unaware off.
